c10/core/impl/cow/state_machine.cpp

```diff
@@ -0,0 +1,240 @@
+
+  // Guards all the state.
+  std::mutex mtx_;
+  // How many writes have been applied to the storage.
+  cow::StateMachine::Generation physical_generation_ = 0;
+  // The shadow storage to use for any tensors that don't have
+  // one. This situation is common, and will be true for tensors and
+  // views thereof created before any copy on writes.
+  cow::ShadowStorageNonIntrusive default_shadow_storage_{physical_generation_};
+};
+
+cow::StateMachine::StateMachine() : state_rep_(StateId::initial) {}
+
+cow::StateMachine::~StateMachine() {
+  // If we created an impl, we are responsible for clenaing this up
+  // here.
+  Rep state_rep = state_rep_;
+  switch (state_rep.as_state_id()) {
+    case StateId::initial:
+      // Nothing to do.
+      break;
+
+    case StateId::initializing:
+      TORCH_INTERNAL_ASSERT(false); // we should not race with the destructor
+
+    case StateId::active:
+      delete &state_rep.as_state_machine_impl();
+  }
+}
+
+auto cow::StateMachine::physical_generation() -> Generation {
+  Impl* impl = maybe_get_impl();
+  return impl != nullptr ? impl->physical_generation_ : 0;
+}
+
+auto cow::StateMachine::maybe_bump(cow::ShadowStorage* maybe_shadow_storage)
+    -> void {
+  Impl* impl = maybe_get_impl();
+  if (impl == nullptr) {
+    // Any created shadow storage should be bound to the physical
+    // storage that it was created from. Hence, there should only be a
+    // shadow storage on a tensor if its own storage created it. We
+    // don't check for the specific matching of the storage and shadow
+    // storage, but we do check that the presence relationship holds.
+    //
+    // TODO: This check should probably be here, but it is actually
+    // triggering in the StaticRuntime.autogen_inner test.
+    //
+    // TORCH_INTERNAL_ASSERT(maybe_shadow_storage == nullptr);
+    return;
+  }
+
+  impl->maybe_bump(maybe_shadow_storage);
+}
+
+auto cow::StateMachine::simulate_lazy_copy(
+    cow::ShadowStorage* maybe_shadow_storage)
+    -> intrusive_ptr<cow::ShadowStorage> {
+  return ensure_initialized().simulate_lazy_copy(maybe_shadow_storage);
+}
+
+auto cow::StateMachine::maybe_get_impl() -> cow::StateMachine::Impl* {
+  return get_impl([] { return nullptr; });
+}
+
+auto cow::StateMachine::ensure_initialized() -> cow::StateMachine::Impl& {
+  return *get_impl([this]() -> c10::optional<cow::StateMachine::Impl*> {
+    Rep initial_rep(StateId::initial);
+    Rep initializing_rep(cow::StateMachine::StateId::initializing);
+    // If we fail to exchange, it means that someone else is
+    // initializing.
+    if (!state_rep_.compare_exchange_strong(initial_rep, initializing_rep)) {
+      return c10::nullopt;
+    }
+
+    auto impl = std::make_unique<cow::StateMachine::Impl>();
+    Rep new_rep(*impl);
+    TORCH_INTERNAL_ASSERT(
+        new_rep.as_state_id() == cow::StateMachine::StateId::active);
+    // We won the race to initializing: we will be the only thread
+    // that may transition to active.
```

Comment:
Instead of spinning on initializing, can't you just speculatively allocate StateMachine::Impl and then throw it out if you lose the write race?

Reply:
This reduces the number of CEX from two to one, and also avoids the need for spinlocking.

Reply:
Good idea and nice subsequent simplification (in progress) that follows.

Reply:
done